### PR TITLE
Add missing lists of valid object type/name in AirLoopHVAC:OutdoorAirSystem:EquipmentList, Branch, PlantEquipmentList, CondenserEquipmentList objects

### PR DESCRIPTION
**Pull request by Nigusse**
Addresses issue #5479 

Added missing reference lists for Object Type and Object Name fields for the following objects:
- AirLoopHVAC:OutdoorAirSystem:EquipmentList
- Branch
- PlantEquipmentList
- CondenserEquipmentList

This branch should not have any diffs.

Pull request overview
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- Defect/Suggestion: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
  - [ ] If new idf included, locally check the err file and other outputs
  - [ ] Documentation changes in place

### CODE REVIEW

**Comment by Nigusse** [comment]
@mjwitte The CI machine has come all green. Unit tests have also been added to this branch and is ready for review.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 15 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
Current thoughts:
- That's a lot of IDD changes.  I don't have a strong opinion there, just stating the obvious.  And you have @mjwitte and @JasonGlazer behind you on the IDD changes, so that's fine.
- The input processor changes could be streamlined to a more generic approach so that if this gets traction, we could add others later (of course JSON......but anyway) without propagating the code to more hardwired pieces.
- The input processor changes weren't needed as a part of this fix.
- Because of the previous two bullet points, and to avoid extra work trying to fix up the input processor changes, my recommendation would be to stash that part of the branch away and get the IDD changes merged in to satisfy this fix.  If you really feel like that needs to go in for this fix, we need to rigor over the code a bit more.

**Comment by Nigusse** [comment]
If there is a better way to streamline the input processor changes, then I do not mind to stash away the code change part.

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -51906,6 +51982,10 @@ AirLoopHVAC:UnitarySystem,
        \required-field
        \type alpha
        \note Unique name for the Unitary System.
+       \reference-class-name validBranchEquipmentTypes
+       \reference validBranchEquipmentNames
+       \reference-class-name validBranchEquipmentTypes
```
Comment (mbadams5) [inline idd/Energy+.idd.in:51987]: @Nigusse copy and paste error

Comment (Nigusse) [inline idd/Energy+.idd.in:51987]: @mbadams5 I will update the idd, merge in develop and will push the branch. Are you working on it?

Comment (mbadams5) [inline idd/Energy+.idd.in:51987]: Working on what? These changes? Or InputProcessor changes in general?

Comment (Nigusse) [inline idd/Energy+.idd.in:51987]: I thought may be you are reviewing it because of the comment.

Comment (mbadams5) [inline idd/Energy+.idd.in:51987]: I just glanced through when I saw Edwin had commented.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 15 days since this pull request was last updated.

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -56166,23 +56262,293 @@ AirLoopHVAC:OutdoorAirSystem:EquipmentList,
        \reference AirLoopOAEquipmentLists
   A2 , \field Component 1 Object Type
        \required-field
+       \type choice
```
Comment (mjwitte) [inline idd/Energy+.idd.in:56265]: @Nigusse Was there a reason that these were not done as \object-list validOAEquipmentTypes?  The IDD can use this approach even if we don't add another set of input processor validation.

File: idd/Energy+.idd.in [hunk 3]
```diff
@@ -67789,6 +68421,8 @@ ThermalStorage:Ice:Simple,
       \type alpha
       \required-field
       \reference IceThermalStorageEquipment
+      \reference-class-name validBranchEquipmentTypes
```
Comment (mjwitte) [inline idd/Energy+.idd.in:68424]: @Nigusse I'm surprised that ice storage objects are not also included in validPlantEquipmentTypes.  I assume you confirmed that in source code?

**Comment by mjwitte** [comment]
There are two other objects that are still missing \object-list references:
`PlantEquipmentOperation:ComponentSetpoint` and `PlantEquipmentOperation:UserDefined`

File: idd/Energy+.idd.in (hunk 2 above)
Comment (Nigusse) [inline idd/Energy+.idd.in:56265]: @mjwitte There was no a specific reason. It could have been added.

File: idd/Energy+.idd.in (hunk 3 above)
Comment (Nigusse) [inline idd/Energy+.idd.in:68424]: @mjwitte If there was a test file that shows the use of "ice storage objects" as a valid plant equipment list object, then the IP would have catch it, and fatal out. That was the case for some of the objects that I missed to include.

**Comment by Nigusse** [comment]
PlantEquipmentOperation:ComponentSetpoint and PlantEquipmentOperation:UserDefined are used in PlantEquipmentOperationSchemes not in PlantEquipmentList object. Should they be added to PlantEquipmentList?

**Comment by mjwitte** [comment]
No, but PlantEquipmentOperation:ComponentSetpoint and PlantEquipmentOperation:UserDefined have fields in them for   Object Type and Object Name that have no key choices or object-list tags.  So, to fulfill the original goal of #5479, we should add that.

**Comment by mjwitte** [comment]
@Nigusse @Myoldmopar I compared the lists of valid OA system and branch objects with code that has explicit if statements to check every valid component type, made a few changes to get the idd to match up.  

Plant and condenser equipment lists are a bit problematic.  The code accepts nearly any valid plant component in a `PlantEquipmentList` even if it can't really supply any heating or cooling.  Components are checked against the valid plant/condenser equipment types declared in [DataPlant](https://github.com/NREL/EnergyPlus/blob/a4a7439ad4a03eedfcd4abe50bbc7584b5ba970e/src/EnergyPlus/DataPlant.cc#L216-L220) and the only other check is that pumps are rejected from equipment lists.  Beyond that, the code lets you put even something like a Pipe on a plant equipment list and it will run without errors.  The attached csv file lists the components and types from DataPlant with an x next to everything currently tagged as valid plant or condenser equipment types.  

Currently, only components that make sense as supply equipment are tagged, but there are some question marks, and I'm having some doubts about this now.  It doesn't make sense to include things like pipes and coils in these lists.  But there may be some additional components that could legitimately be on a plant equipment list.  So, I backed off on the object list checks in the input processor so it's just a severe but does not terminate the run.

[PlantEquipment Types-csv.txt](https://github.com/NREL/EnergyPlus/files/425191/PlantEquipment.Types-csv.txt)

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@Nigusse @lgentile it has been 14 days since this pull request was last updated.

File: idd/Energy+.idd.in [hunk 4]
```diff
@@ -34259,6 +34261,8 @@ ZoneHVAC:Baseboard:RadiantConvective:Water,
        \min-fields 12
   A1,  \field Name
        \required-field
+       \reference-class-name validBranchEquipmentTypes
```
Comment (mjwitte) [inline idd/Energy+.idd.in:34264]: @mbadams5 Let's use ZoneHVAC:BaseBoard:RadiantConvective:Water as a discussion example.
\reference-class-name validBranchEquipmentTypes adds "ZoneHVAC:BaseBoard:RadiantConvective:Water" to the list validBranchEquipmentTypes 
This list is static and doesn't change based on the presence or number of baseboards in an idf file
\reference validBranchEquipmentNames adds the\* names\* of all ZoneHVAC:BaseBoard:RadiantConvective:Water objects to the list validBranchEquipmentNames 
This list is variable and includes all the names of all baseboards (and other valid branch equipment) that are present in the idf.

File: idd/Energy+.idd.in [hunk 5]
```diff
@@ -58441,73 +58631,117 @@ Branch,
         \note Air loops do not account for pressure drop using this field
         \note Valid curve types are: Curve:Functional:PressureDrop or
         \note one of Curve:{Linear,Quadratic,Cubic,Exponent}')
-   A3, \field Component 1 Object Type
+   A3,  \field Component 1 Object Type
         \begin-extensible
         \required-field
-   A4, \field Component 1 Name
+        \type object-list 
+        \object-list validBranchEquipmentTypes
```
Comment (mjwitte) [inline idd/Energy+.idd.in:58638]: In the branch object, Component n Object Type and Component n Object Name are currently free form fields with no indication of what types or names are valid.  This adds \object-list validBranchEquipmentTypes to the Component n Object Type field to indicate what object types are valid.  ZoneHVAC:BaseBoard:RadiantConvective:Water will be part of this list.  The alternative is to make this a choice field and add a list of explicit keys for every valid branch equipement type.
\type choice
\key ZoneHVAC:BaseBoard:RadiantConvective:Water
\key Fan:VariableVolume
\key Fan:ConstantVolume
etc.
And then repeat these keys on every field down the line.  The long list of hard keys is used elsewhere in the IDD, but it's messy.  See ZoneHVAC:EquipmentList, for example.

```
ZoneHVAC:EquipmentList,
  A1 , \field Name
       \required-field
       \reference ZoneEquipmentLists
  A2 , \field Zone Equipment 1 Object Type
       \begin-extensible
       \type choice
       \key ZoneHVAC:TerminalUnit:VariableRefrigerantFlow
       \key ZoneHVAC:AirDistributionUnit
       \key AirTerminal:SingleDuct:Uncontrolled
       \key ZoneHVAC:EnergyRecoveryVentilator
       \key ZoneHVAC:EvaporativeCoolerUnit
       \key ZoneHVAC:FourPipeFanCoil
       \key ZoneHVAC:OutdoorAirUnit
       \key ZoneHVAC:PackagedTerminalAirConditioner
       \key ZoneHVAC:PackagedTerminalHeatPump
       \key ZoneHVAC:UnitHeater
       \key ZoneHVAC:UnitVentilator
       \key ZoneHVAC:VentilatedSlab
       \key ZoneHVAC:WaterToAirHeatPump
       \key ZoneHVAC:WindowAirConditioner
       \key ZoneHVAC:Baseboard:RadiantConvective:Electric
       \key ZoneHVAC:Baseboard:RadiantConvective:Water
       \key ZoneHVAC:Baseboard:RadiantConvective:Steam
       \key ZoneHVAC:CoolingPanel:RadiantConvective:Water
       \key ZoneHVAC:Baseboard:Convective:Electric
       \key ZoneHVAC:Baseboard:Convective:Water
       \key ZoneHVAC:HighTemperatureRadiant
       \key ZoneHVAC:LowTemperatureRadiant:VariableFlow
       \key ZoneHVAC:LowTemperatureRadiant:ConstantFlow
       \key ZoneHVAC:LowTemperatureRadiant:Electric
       \key ZoneHVAC:Dehumidifier:DX
       \key ZoneHVAC:IdealLoadsAirSystem
       \key ZoneHVAC:RefrigerationChillerSet
       \key Fan:ZoneExhaust
       \key WaterHeater:HeatPump:PumpedCondenser
       \key WaterHeater:HeatPump:WrappedCondenser
       \required-field
  A3 , \field Zone Equipment 1 Name
       \required-field
```

**Comment by Myoldmopar** [comment]
OK, I pulled all code changes out, leaving only the idd changes.  I'll make sure CI is OK with the changes, and @mjwitte, maybe you could run some tests in IDF Editor land?

**Review by Myoldmopar — APPROVED** [review]
OK, this has been whittled down to just the idd changes.  I'm OK with these changes.  @mjwitte can you test it in IDF editor?  If not, I can do it in a VM.  Thanks!

**Comment by mjwitte** [comment]
@Myoldmopar Edited the title and description.  Tested revised IDD with IDF Editor.  All looks good.  Merge at will.

**Pull request merged**